src/main/java/iterator/RichIterator.java

```diff
@@ -326,7 +534,31 @@ default <B> B foldLeft(B zero, BiFunction<? super B, ? super A, ? extends B> acc
      * Iterator(1,2,3,4,5).scanLeft(0, Integer::sum) // Iterator(0,1,3,6,10,15)
      */
     default <B> RichIterator<B> scanLeft(B zero, BiFunction<? super B, ? super A, ? extends B> acc) {
-        throw new NotImplementedException();
+        return new RichIterator<B>() {
+            private B result;
+
+            @Override
+            public boolean hasNext() {
+                if (result == null) {
+                    return true;
+                } else {
+                    return RichIterator.this.hasNext();
+                }
+            }
+
+            @Override
+            public B next() {
+                if (result == null) {
+                    result = zero;
+                } else if (hasNext()) {
+                    result = acc.apply(result, RichIterator.this.next());
+                } else {
+                    throw new NoSuchElementException();
+                }
+
+                return result;
+            }
+        };
```

Comment:
You can avoid all the checks for first element by simply ignoring it and prepend it to the iterator (also relevant for `iterate`)
```suggestion
        }.prepend(zero);
```